File: kotlin-node/karakum/src/jsMain/resources/injections/ModuleInjection.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

function isModuleInterfaceNode(node) {
    return (
        ts.isInterfaceDeclaration(node)
        && node.name.text === "Module"

        && node.parent
        && ts.isModuleBlock(node.parent)

        && node.parent.parent
        && ts.isModuleDeclaration(node.parent.parent)
        && ts.isIdentifier(node.parent.parent.name)
        && node.parent.parent.name.text === "NodeJS"
    )
}

export default {
    setup(context) {
        this.moduleInterfaceNode = null
    },

    traverse(node) {
        if (isModuleInterfaceNode(node)) {
            this.moduleInterfaceNode = node
        }
    },

    render(node, context, next) {
        const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

        if (
            sourceFileName.endsWith("module.d.ts")
            && ts.isHeritageClause(node)

            && node.parent
            && ts.isInterfaceDeclaration(node.parent)
            && node.parent.name.text === "Module"
        ) {
            return ""
        }

        return null
    },

    inject(node, context, render) {
        const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

        if (
            sourceFileName.endsWith("module.d.ts")
            && ts.isClassDeclaration(node)
            && node?.name.text === "Module"
            && context.type === "MEMBER"
        ) {
            return this.moduleInterfaceNode.members
                .map(member => render(member))
        }

        return []
    },

    generate(context, render) {
        return []
    }
}

File: kotlin-node/karakum/src/jsMain/kotlin/wrappersgenerator/node/injections/ModuleInjection.kt
package wrappersgenerator.node.injections

import arrow.core.raise.impure
import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.*
import io.github.sgrishchenko.karakum.util.getParentOrNull
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import typescript.*

private fun isModuleInterfaceNode(node: Node) = nullable {
    ensure(isInterfaceDeclaration(node))
    ensure(node.name.text == "Module")

    val moduleBlock = ensureNotNull(node.getParentOrNull())
    ensure(isModuleBlock(moduleBlock))

    val module = ensureNotNull(moduleBlock.getParentOrNull())
    ensure(isModuleDeclaration(module))

    val moduleName = module.name
    ensure(isIdentifier(moduleName))
    ensure(moduleName.text == "NodeJS")
} != null

class ModuleInjection : Injection {
    private lateinit var moduleInterfaceNode: InterfaceDeclaration

    override fun setup(context: Context) = Unit

    override fun traverse(node: Node, context: Context) = impure {
        ensure(isInterfaceDeclaration(node))
        ensure(isModuleInterfaceNode(node))

        moduleInterfaceNode = node
    }

    override fun render(node: Node, context: Context, next: Render<Node>) = nullable {
        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        ensure(sourceFileName.endsWith("module.d.ts"))

        ensure(isHeritageClause(node))

        val interfaceNode = ensureNotNull(node.getParentOrNull())
        ensure(isInterfaceDeclaration(interfaceNode))
        ensure(interfaceNode.name.text == "Module")

        ""
    }

    override fun inject(node: Node, context: InjectionContext, render: Render<Node>) = nullable {
        ensure(context.type == InjectionType.MEMBER)

        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        ensure(sourceFileName.endsWith("module.d.ts"))

        ensure(isClassDeclaration(node))
        ensure(node.name?.text == "Module")

        moduleInterfaceNode.members.asArray()
            .map { render(it) }
            .toTypedArray()
    }

    override fun generate(context: Context, render: Render<Node>) = emptyArray<GeneratedFile>()
}
